R/pool.R

```diff
@@ -19,164 +18,238 @@ NULL
-      canary <- new.env(parent = emptyenv())
-      canary$closed <- FALSE
-      attr(object, "canary") <- canary
-      ## maybe calling gc() ins't necessary (see issue #3 in Github)
-      gc()
-      reg.finalizer(canary, function(e) {
-        protectDefaultScheduler({
-          if (!canary$closed) {
-            ## for dplyr use (as of right now), leaking is inevitable
-            ## and it will annoy the heck out of users to be getting
-            ## warnings when they're doing exactly as we tell them to
-            ## so commenting this out for now...
-            # warning("You have leaked pooled objects. Closing them.")
+      if (is.null(object)) {
+        stop("Object creation was not successful. The `factory` ",
+             "argument must be a function that creates and ",
+             "returns the object to be pooled.")
+      }
+
+      ## attach metadata about the object
+      pool_metadata <- new.env(parent = emptyenv())
+      attr(object, "pool_metadata") <- pool_metadata
+
+      id <- as.character(private$idCounter)
+      private$idCounter <- private$idCounter + 1
+      pool_metadata$id <- id
+      pool_metadata$pool <- self
+      pool_metadata$valid <- TRUE
+      pool_metadata$state <- "free"
+      pool_metadata$lastValidated <- NULL
+
+      ## detect leaked connections and destroy them
+      reg.finalizer(pool_metadata, function(e) {
+        naiveScheduler$protect({
+          if (pool_metadata$valid) {
+            warning("You have a leaked pooled object. Destroying it.")
             scheduleTask(1, function() {
-              ## changed because of issue #4 in Github, I really think
-              ## this makes more sense...
-              #self$release(id, object)
-              #message("Connection finalizer ran")
-
-              ## changed back to this, pending discussion with @jcheng5
-              private$changeObjectStatus(id, object, "taken", NULL)
-              private$destroyObject(object)
+              state <- attr(object, "pool_metadata", exact = TRUE)$state
+              private$changeObjectStatus(id, object, state, NULL)
             })
           }
         })
-      })
+      },
+      onexit = TRUE)
 
       private$changeObjectStatus(id, object, NULL, "free")
       return(object)
     },
+
+    ## tries to run onDestroy
     destroyObject = function(object) {
-      canary <- attr(object, "canary", exact = TRUE)
-      if (canary$closed) {
-        warning("Object was destroyed twice")
-      }
-      canary$closed <- TRUE
-      onDestroy(object)
+      tryCatch({
+        pool_metadata <- attr(object, "pool_metadata", exact = TRUE)
+        if (!pool_metadata$valid) {
+          warning("Object was destroyed twice.")
+          return()
+        }
+        pool_metadata$valid <- FALSE
+        private$cancelScheduledTask(object, "validateHandle")
+        onDestroy(object)
+      }, error = function(e) {
+        warning("Object of class ", is(object)[1],
```

Comment:
I don't care either way, I just had to idea you could even use `is` with a single argument.